enhancements/machine-config/on-cluster-layering.md

```diff
@@ -0,0 +1,1137 @@
+There are situations where the on-cluster build will fail. For example, if a
+cluster admin's custom Containerfile is dynamically inferring the OS version from
+the `/etc/os-release` file to ensure they set up the correct RPM repository
+path, this could break when the OS version changes. Another case to consider is
+whenever a cluster admin has overridden their `osImageURL` value with a
+custom-built image produced by an off-cluster build system; though this scenario
+is not solely an upgrade concern.
+
+For this reason, we should consider a way to allow a cluster admin to perform
+pre-upgrade checks such as performing a trial build before committing to the
+upgrade process. This way, they can adjust their custom Containerfile content and
+pre-build the images prior to beginning the upgrade process.
+
+In addition to this, a cluster admin in a cloud environment should be able to
+bring up an ephemeral test node with the latest OS on it, verify that it can
+connect to the Kube API server, receive workloads, etc. For bare-metal cases, a
+cluster admin could temporarily repurpose one of their worker nodes for this
+purpose. We could eventually provide some automation around this process.
+
+These checks could also be applicable to general config rollouts as a form of
+preflight check.
+
+#### Entitlements
+    
+There was significant engineering work put into Podman and Buildah to enable
+seamless consumption of RHEL-entitled content. OpenShift has a similar model
+for entitled clusters, with the `etc-pki-entitlement` secret being present
+in the `openshift-config-managed` namespace. Given this, it is possible that
+we can consume RHEL-entitled content within the OCL build process itself,
+with a few workarounds:
+    
+- We need to clone the `etc-pki-entitlement` certificates into the MCO
+  namespace.
+- These need to be mounted into the build pod.
+- These need to be copied from the secret volume mount into a place where
+  Buildah can consume them.
+
+In the future, we could make use of the [shared resource CSI driver](https://docs.openshift.com/container-platform/4.17/storage/container_storage_interface/ephemeral-storage-shared-resource-csi-driver-operator.html)
+to automatically keep this secret up-to-date. In lieu of that, the cluster
+admin will need to keep the secret up-to-date or the MCO will have to take
+on that role as well.
+    
+### Risks and Mitigations
+
+What are the risks of this proposal and how do we mitigate. Think
+broadly. For example, consider both security and how this will impact
+the larger OKD ecosystem.
+
+How will security be reviewed and by whom?
+
+- For the custom pod builder, a review by someone who is knowledgeable
+  about container builders, RBAC, and SELinux would be helpful to ensure
+  that we can enable this behavior while avoiding the need for elevated
+  permissions / privileges.
+
+How will UX be reviewed and by whom?
+
+- TBD.
+
+### Drawbacks
+
+TBD.
+
+## Design Details
+
+### Open Questions
```

Comment:
There's quite a few open questions here, should we be answering some of these prior to merging?